x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/PriorityProcessWorkerExecutorService.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.pytorch;
+
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.SuppressForbidden;
+import org.elasticsearch.xpack.ml.job.process.AbstractProcessWorkerExecutorService;
+
+import java.util.concurrent.PriorityBlockingQueue;
+
+/**
+ * An {@link AbstractProcessWorkerExecutorService} where the runnables
+ * are executed in priority order with a tie breaker value for requests
+ * of equal priority. The tie breaker can be used to preserve insertion order.
+ *
+ * Order is maintained by a PriorityQueue
+ */
+public class PriorityProcessWorkerExecutorService extends AbstractProcessWorkerExecutorService<
+    PriorityProcessWorkerExecutorService.OrderedRunnable> {
+
+    public enum RequestPriority {
+        HIGHEST,
+        HIGH,
+        NORMAL
+    };
+
+    /**
+     * A Runnable sorted first by RequestPriority then a tie breaker which in
+     * most cases will be the insertion order
+     */
+    public static record OrderedRunnable(RequestPriority priority, long tieBreaker, Runnable runnable)
+        implements
+            Comparable<OrderedRunnable>,
+            Runnable {
+        @Override
+        public int compareTo(OrderedRunnable o) {
+            int p = this.priority.compareTo(o.priority);
+            if (p == 0) {
+                return (int) (this.tieBreaker - o.tieBreaker);
+            }
+
+            return p;
+        }
+
+        @Override
+        public void run() {
+            runnable.run();
+        }
+    };
+
+    private final int queueCapacity;
+
+    /**
+     * @param contextHolder the thread context holder
+     * @param processName the name of the process to be used in logging
+     * @param queueCapacity the capacity of the queue holding operations. If an operation is added
+     *                  for execution when the queue is full a 429 error is thrown.
+     */
+    @SuppressForbidden(reason = "properly rethrowing errors, see EsExecutors.rethrowErrors")
+    public PriorityProcessWorkerExecutorService(ThreadContext contextHolder, String processName, int queueCapacity) {
+        super(contextHolder, processName, queueCapacity, PriorityBlockingQueue::new);
```

Comment:
The other option here is to create a queue that wraps three separate queues (normal, high, highest). It would take a "prioritized" object and feed it to the appropriate queue. Poll, etc. would then pull from the appropriate queue and it would keep track of its size as a sum of the inner queues. 

I am not sure if that is any cleaner/faster or nicer than simply using a priority queue like you got here. But it would prevent having to reshuffle the order on offer every time.

Reply:
@davidkyle I dont think you have to sit there and poll like that. I would poll from all three, if its null, wait on a lock condition that is signaled when something is added to your queue wrapper. Let me see if I can write up a quick example...

```java
static class MyQueues<E> implements BlockingQueue<E> {

        private final Queue<E> queueNormal;
        private final Queue<E> queueHigh;
        private final Queue<E> queueHighest;
        private final int sizeLimit;

        // lock for queueing
        private final ReentrantLock lock = new ReentrantLock();

        // For waiting to not be empty
        private final Condition notEmpty = lock.newCondition();

        MyQueues(int sizeLimit) {
            queueNormal = new ArrayDeque<>();
            queueHigh = new ArrayDeque<>();
            queueHighest = new ArrayDeque<>();
            this.sizeLimit = sizeLimit;
        }

        public boolean offerWithPriority(E item, RequestPriority priority) {
            lock.lock();
            try {
                if (queueNormal.size() + queueHigh.size() + queueHighest.size() > sizeLimit) {
                    return false;
                }
                switch (priority) {
                    case HIGH -> queueHigh.add(item);
                    case NORMAL -> queueNormal.add(item);
                    case HIGHEST -> queueHighest.add(item);
                }
                notEmpty.signal();
                return true;
            } finally {
                lock.unlock();
            }
        }

        @Override
        public E poll(long timeout, TimeUnit unit) throws InterruptedException {
            long nanos = unit.toNanos(timeout);
            lock.lockInterruptibly();
            try {
                E item = null;
                while ((item = pollUnsafe()) == null && nanos > 0) {
                    nanos = notEmpty.awaitNanos(nanos);
                }
                return item;
            } finally {
                lock.unlock();
            }
        }

        @Override
        public E poll() {
            lock.lock();
            try {
                return pollUnsafe();
            } finally {
                lock.unlock();
            }
        }

        private E pollUnsafe() {
            return Optional.ofNullable(queueHighest.poll())
                .or(() -> Optional.ofNullable(queueHigh.poll()))
                .or(() -> Optional.ofNullable(queueNormal.poll()))
                .orElse(null);
        }
// <snip> all the other methods for blocking queue </snip>
    }
```

Reply:
++ looks good I think that would work.

I paused before implementing this- as usual the best thing to do is run a micro benchmark first. 

The concern is that offers into the priority queue will be slow. For the common use case we expect requests to have normal priority and be naturally ordered by the request Id which does not cause a reshuffling of the priority queue. 

For this case I compared offering 1000 items in their natural order into a `LinkedBlockingQueue` and `PriorityBlockingQueue`. Surprisingly the priority queue was faster 17807.750 ns vs 26797.762 ns on average.

Inserting 1000 items into the priority queue where every 100th item has a higher priority was still faster then inserting into the linked queue at 20959.939 ns (average)